hooks/shotgun_launch_publish.py

```diff
@@ -77,37 +78,119 @@ def execute(self, path, context, associated_entity, **kwargs):
-        
         # in older configs, launch instances were named tk-shotgun-launchmaya
         # in newer configs, launch instances are named tk-multi-launchamaya
         old_config = "tk-shotgun-%s" % launch_app_instance_name
         new_config = "tk-multi-%s" % launch_app_instance_name
+        app_instance = None
         
         if old_config in self.parent.engine.apps:
             # we have a tk-shotgun-xxx instance in the config
             app_instance = old_config
-        
         elif new_config in self.parent.engine.apps:
             # we have a tk-multi-xxx instance in the config
             app_instance = new_config
-            
-        else:
-            raise TankError("The '%s' app could not be found in the '%s' "
-                            "environment!" % (new_config, self.parent.engine.environment.get("name")))        
-        
-        
+
+        return app_instance
+
+    def _do_launch(self, launch_app_instance_name, engine_name, path, context):
+        """
+        Tries to create folders then launch the publish.
+        """
         # first create folders based on the context - this is important because we 
         # are creating them in deferred mode, meaning that in some cases, new user sandboxes
         # maybe created at this point.
         if context.task:
             self.parent.tank.create_filesystem_structure("Task", context.task["id"], engine_name)
         elif context.entity:
             self.parent.tank.create_filesystem_structure(context.entity["type"], context.entity["id"], engine_name)
+        
+        # in ancient configs, launch instances were named tk-shotgun-launchmaya
+        # in less-ancient configs, launch instances are named tk-multi-launchamaya
+        old_config = "tk-shotgun-%s" % launch_app_instance_name
+        new_config = "tk-multi-%s" % launch_app_instance_name
+        app_instance = self._get_legacy_launch_command(launch_app_instance_name)
+
+        # If we didn't find anything useful in the current context, then we
+        # can check the target context. This type of configuration is the approach
+        # taken in modern configurations based on tk-config-default2 for secondary
+        # entity types like PublishedFile and Version, hence it's likely that this
+        # approach will get us where we need to go.
+        if app_instance is None:
+            # Changing context here allows us to find the launchapp that is
+            # configured in the target context rather than requiring configuration
+            # for every engine in the source context. This is a result of launchapp's
+            # need of engine configurations to be present for the Software launcher
+            # functionality to be used.
+            sgtk.platform.change_context(context)
+
+            # One last chance to find a legacy-style launcher.
+            app_instance = self._get_legacy_launch_command(launch_app_instance_name)
+
+            if app_instance is None:
```

Comment:
minor: suggest swapping this conditional around to avoid the intermediate `return` 

```python
if app_instance:
    # now try to lauch this via the tk-multi-launchapp
    ...
else:
    # We're likely in a situation where Software-entity launchers are
    ...
```